drivers/nec/display/display.cr

```diff
@@ -0,0 +1,397 @@
+      end
+    end
+
+    task.try &.success
+  end
+
+  def do_poll
+    power?(priority: 0)
+    logger.debug { "Polling, power = #{self[:power]}" }
+
+    if self[:power]?
+      power_on_delay
+      mute_status
+      volume_status
+      video_input
+      audio_input
+    end
+  end
+
+  private def parse_response(data : String)
+    # 14..15 == type (we don't care)
+    max = data[16..19].to_i(16)
+    value = data[20..23].to_i(16)
+
+    case Command.from_value(data[10..13].to_i(16))
+    when .video_input?
+      input = Inputs.from_value(value)
+      self[:input] = input
+      target_input = @target_input ||= input
+      if target_input && self[:input] != self[:target_input]
+        switch_to(target_input)
+      end
+    when .audio_input?
+      self[:audio] = Audio.from_value(value)
+      target_audio = @target_audio
+      if target_audio && self[:audio] != self[:target_audio]
+        switch_audio(target_audio)
+      end
+    when .volume_status?
+      self[:volume_max] = max
+      unless self[:audio_mute]
+        self[:volume] = value
+      end
+    when .brightness_status?
+      self[:brightness_max] = max
+      self[:brightness] = value
+    when .contrast_status?
+      self[:contrast_max] = max
+      self[:contrast] = value
+    when .mute_status?
+      self[:audio_mute] = value == 1
+      if(value == 1)
+        self[:volume] = 0
+      else
+        volume_status(60) # high priority
+      end
+    when .power_on_delay?
+      if value > 0
+        self[:warming] = true
+        schedule.in(value.seconds) do # Prevent any commands being sent until the power on delay is complete
+          power_on_delay.as(PlaceOS::Driver::Transport)
+        end
+      else
+        schedule.in(3.seconds) do # Reactive the interface once the display is online
+          self[:warming] = false # allow access to the display
+        end
+      end
+    when .auto_setup?
+      # auto_setup
+      # nothing needed to do here (we are delaying the next command by 4 seconds)
+    else
+      logger.info { "-- NEC LCD, unknown response: #{data[10..13]}" }
+      logger.info { "-- NEC LCD, full response was: #{data}" }
+    end
+  end
+
+  # Types of messages sent to and from the LCD
+  enum MsgType
+    Command             = 0x41 # 'A'
+    Command_reply       = 0x42 # 'B'
+    Get_parameter       = 0x43 # 'C'
+    Get_parameter_reply = 0x44 # 'D'
+    Set_parameter       = 0x45 # 'E'
+    Set_parameter_reply = 0x46 # 'F'
+  end
+
+  enum Command
+    Video_input       = 0x0060
+    Audio_input       = 0x022E
+    Volume_status     = 0x0062
+    Mute_status       = 0x008D
+    Power_on_delay    = 0x02D8
+    Contrast_status   = 0x0012
+    Brightness_status = 0x0010
+    Auto_setup        = 0x001E
+    Power_query       = 0x01D6
+
+    def to_s : String
+      Nec::Display::All.format_value(self.value)
+    end
+  end
+
+  def self.format_value(value : Int, length : Int = 4) : String
+    value.to_s(16).upcase.rjust(length, '0')
```

Comment:
[`String#to_s`](https://crystal-lang.org/api/0.35.1/Int.html#to_s(base:Int=10,*,upcase:Bool=false):String-instance-method) provides an `upcase` param to save a double string creation here.